File: tests/cpp/test_audio_initialization.cpp
/**
 * Audio System Initialization Tests
 * 
 * Tests system initialization, shutdown, AudioSession, and logging controls.
 */

#include "test_common.h"
#include "audio_manager.h"
#include "audio_session.h"
#include "logging.h"
#include <iostream>

using namespace audio;

void test_initialization() {
    TEST("System Initialization")
    
    auto& audio = AudioManager::GetInstance();
    
    // System should already be initialized from main()
    ASSERT(!audio.Initialize(), "Initialize should return false when already initialized")
    
    // Shutdown and reinitialize
    audio.Shutdown();
    ASSERT(audio.Initialize(), "Should be able to reinitialize after shutdown")
    
    END_TEST
}

void test_audio_session_usage() {
    TEST("AudioSession Usage")
    
    auto& audio = AudioManager::GetInstance();
    audio.Shutdown();
    
    {
        AudioSession session;
        GroupHandle group = audio.CreateGroup("session_group");
        ASSERT(group.IsValid(), "AudioSession should allow group creation")
        audio.SetGroupVolume(group, 0.5f);
        audio.DestroyGroup(group);
    }
    
    ASSERT(audio.Initialize(), "AudioManager should reinitialize after AudioSession scope")
    
    END_TEST
}

void test_logging_controls() {
    TEST("Logging Controls")
    
    // Test that logging is always available (no compile-time flag needed)
    auto original = AudioManager::GetLogLevel();
    
    // Test all log levels can be set
    AudioManager::SetLogLevel(LogLevel::Off);
    ASSERT(AudioManager::GetLogLevel() == LogLevel::Off, "SetLogLevel should update to Off")
    ASSERT(!Logger::IsEnabled(LogLevel::Error), "Nothing should be enabled at Off level")
    
    AudioManager::SetLogLevel(LogLevel::Error);
    ASSERT(AudioManager::GetLogLevel() == LogLevel::Error, "SetLogLevel should update to Error")
    ASSERT(Logger::IsEnabled(LogLevel::Error), "Error should be enabled at Error level")
    ASSERT(!Logger::IsEnabled(LogLevel::Warn), "Warn should be disabled at Error level")
    
    AudioManager::SetLogLevel(LogLevel::Warn);
    ASSERT(AudioManager::GetLogLevel() == LogLevel::Warn, "SetLogLevel should update to Warn")
    ASSERT(Logger::IsEnabled(LogLevel::Error), "Error should be enabled at Warn level")
    ASSERT(Logger::IsEnabled(LogLevel::Warn), "Warn should be enabled at Warn level")
    ASSERT(!Logger::IsEnabled(LogLevel::Info), "Info should be disabled at Warn level")
    
    AudioManager::SetLogLevel(LogLevel::Info);
    ASSERT(AudioManager::GetLogLevel() == LogLevel::Info, "SetLogLevel should update to Info")
    ASSERT(Logger::IsEnabled(LogLevel::Error), "Error should be enabled at Info level")
    ASSERT(Logger::IsEnabled(LogLevel::Warn), "Warn should be enabled at Info level")
    ASSERT(Logger::IsEnabled(LogLevel::Info), "Info should be enabled at Info level")
    ASSERT(!Logger::IsEnabled(LogLevel::Debug), "Debug should be disabled at Info level")
    
    AudioManager::SetLogLevel(LogLevel::Debug);
    ASSERT(AudioManager::GetLogLevel() == LogLevel::Debug, "SetLogLevel should update to Debug")
    ASSERT(Logger::IsEnabled(LogLevel::Info), "Info should be enabled at Debug level")
    ASSERT(Logger::IsEnabled(LogLevel::Debug), "Debug should be enabled at Debug level")
    
    // Restore original level
    AudioManager::SetLogLevel(original);
    END_TEST
}

void test_rapid_shutdown_reinitialize() {
    TEST("Rapid Shutdown/Reinitialize Cycles")
    
    auto& audio = AudioManager::GetInstance();
    
    // Perform multiple shutdown/reinitialize cycles
    for (int i = 0; i < 5; i++) {
        audio.Shutdown();
        wait_ms(50);
        ASSERT(audio.Initialize(), "Should be able to reinitialize after shutdown")
        wait_ms(50);
        
        // Verify system works after reinitialize
        GroupHandle group = audio.CreateGroup("test");
        audio.SetGroupVolume(group, 0.5f);
        ASSERT(std::abs(audio.GetGroupVolume(group) - 0.5f) < 0.01f, 
               "System should work after reinitialize")
        audio.DestroyGroup(group);
    }
    
    END_TEST
}

int main(int argc, char* argv[]) {
    std::cout << "========================================" << std::endl;
    std::cout << "Audio System Initialization Tests" << std::endl;
    std::cout << "========================================" << std::endl;
    
    sound_dir = get_sound_dir(argc, argv);
    std::cout << "Sound directory: " << sound_dir << std::endl;
    
    // Initialize audio system
    auto& audio = AudioManager::GetInstance();
    if (!audio.Initialize()) {
        std::cerr << "FATAL: Failed to initialize audio system" << std::endl;
        return 1;
    }
    
    // Run all tests
    test_initialization();
    test_audio_session_usage();
    test_logging_controls();
    test_rapid_shutdown_reinitialize();
    
    // Final shutdown
    audio.Shutdown();
    
    // Print results
    std::cout << "\n========================================" << std::endl;
    std::cout << "Test Results" << std::endl;
    std::cout << "========================================" << std::endl;
    std::cout << "Passed: " << tests_passed << std::endl;
    std::cout << "Failed: " << tests_failed << std::endl;
    std::cout << "Total:  " << (tests_passed + tests_failed) << std::endl;
    
    if (tests_failed == 0) {
        std::cout << "\nALL TESTS PASSED ✓" << std::endl;
        return 0;
    } else {
        std::cout << "\nSOME TESTS FAILED ✗" << std::endl;
        return 1;
    }
}

File: tests/python/test_audio_initialization.py
"""
Audio System Initialization Tests
Tests system initialization, shutdown, AudioSession, and logging controls.
"""

from test_common import *

def test_basic_initialization():
    """Test: Basic system initialization and shutdown"""
    print("Test: Basic initialization... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    assert audio.initialize(), "Failed to initialize"
    audio.shutdown()
    print("PASS")

def test_audio_session_usage():
    """Test: AudioSession usage for initialization/shutdown"""
    print("Test: AudioSession usage... ", end="", flush=True)
    session = game_audio.AudioSession()
    audio = game_audio.AudioManager.get_instance()
    group = audio.create_group("session_test")
    audio.set_group_volume(group, 0.5)
    audio.destroy_group(group)
    session.close()
    print("PASS")

def test_logging_controls():
    """Test: Logging controls are accessible and work correctly"""
    print("Test: Logging controls... ", end="", flush=True)
    
    # Test that logging is always available (no compile-time flag needed)
    original = game_audio.AudioManager.get_log_level()
    
    # Test all log levels can be set
    game_audio.AudioManager.set_log_level(game_audio.LogLevel.Off)
    assert game_audio.AudioManager.get_log_level() == game_audio.LogLevel.Off, "Log level should be Off"
    
    game_audio.AudioManager.set_log_level(game_audio.LogLevel.Error)
    assert game_audio.AudioManager.get_log_level() == game_audio.LogLevel.Error, "Log level should be Error"
    
    game_audio.AudioManager.set_log_level(game_audio.LogLevel.Warn)
    assert game_audio.AudioManager.get_log_level() == game_audio.LogLevel.Warn, "Log level should be Warn"
    
    game_audio.AudioManager.set_log_level(game_audio.LogLevel.Info)
    assert game_audio.AudioManager.get_log_level() == game_audio.LogLevel.Info, "Log level should be Info"
    
    game_audio.AudioManager.set_log_level(game_audio.LogLevel.Debug)
    assert game_audio.AudioManager.get_log_level() == game_audio.LogLevel.Debug, "Log level should be Debug"
    
    # Restore original level
    game_audio.AudioManager.set_log_level(original)
    print("PASS")

def test_rapid_shutdown_reinitialize():
    """Test: Rapid shutdown/reinitialize cycles"""
    print("Test: Rapid shutdown/reinitialize... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    
    for i in range(5):
        audio.shutdown()
        wait_ms(50)
        assert audio.initialize(), f"Should be able to reinitialize after shutdown (cycle {i+1})"
        wait_ms(50)
        
        # Verify system works after reinitialize
        group = audio.create_group("test")
        audio.set_group_volume(group, 0.5)
        assert abs(audio.get_group_volume(group) - 0.5) < 0.01, "System should work after reinitialize"
        audio.destroy_group(group)
    
    # Ensure we shutdown at the end
    audio.shutdown()
    print("PASS")

def run_all_tests():
    """Run all initialization tests"""
    print("=== Python Audio Initialization Tests ===\n")
    
    tests = [
        test_basic_initialization,
        test_audio_session_usage,
        test_logging_controls,
        test_rapid_shutdown_reinitialize,
    ]
    
    passed = 0
    failed = 0
    
    for test in tests:
        try:
            test()
            passed += 1
        except AssertionError as e:
            print(f"FAIL - {e}")
            failed += 1
        except Exception as e:
            print(f"ERROR - {e}")
            failed += 1
    
    print(f"\n=== Results: {passed} passed, {failed} failed ===")
    return failed == 0

if __name__ == "__main__":
    success = run_all_tests()
    sys.exit(0 if success else 1)
